onnxruntime/core/providers/webnn/builders/impl/unsqueeze_op_builder.cc

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Intel Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/common.h"
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/webnn/builders/helper.h"
+#include "core/providers/webnn/builders/model_builder.h"
+#include "core/providers/webnn/builders/op_builder_factory.h"
+
+#include "core/optimizer/initializer.h"
+
+#include "base_op_builder.h"
+#include "builder_utils.h"
+
+namespace onnxruntime {
+namespace webnn {
+
+class UnsqueezeOpBuilder : public BaseOpBuilder {
+  // Add operator related.
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related.
+ private:
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                         const logging::Logger& logger) const override;
+};
+
+// Add operator related.
+void UnsqueezeOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  // Unsqueeze opset 13 uses input 1 as axes, add it to initializer skip list.
+  const auto& input_defs = node.InputDefs();
+  if (node.SinceVersion() > 12 && input_defs.size() > 1) {
+    model_builder.AddInitializerToSkip(input_defs[1]->Name());  // "axes"
+    model_builder.AddInputToSkip(input_defs[1]->Name());
+  }
+}
+
+// Add operator related.
+
+Status UnsqueezeOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                                 const Node& node,
+                                                 const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  emscripten::val input = model_builder.GetOperand(input_defs[0]->Name());
+  std::vector<int64_t> input_shape;
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[0], input_shape, logger), "Cannot get input shape");
+  const auto input_rank = input_shape.size();
+
+  NodeAttrHelper helper(node);
+  emscripten::val options = emscripten::val::object();
+  std::vector<int32_t> axes_data;
+
+  if (node.SinceVersion() >= 13) {
+    // Input axes is provided, use axes initializer data.
+    const auto& initializers(model_builder.GetInitializerTensors());
+    const auto& axes_tensor = *initializers.at(input_defs[1]->Name());
+    Initializer axes_initializer(axes_tensor);
+    const auto axes_data_span = axes_initializer.DataAsSpan<int64_t>();
```

Comment:
Does `DataAsSpan` validate the tensor type is actually int64_t first before casting? If not, we should validate before accessing the memory:

e.g.
`ORT_RETURN_IF_NOT(axes_initializer.data_type() != ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64, "Unsqueeze axes should be int64 data type.");`

Or maybe what you did in `SliceOpBuilder::IsOpSupportedImpl`:

```
if (shape_tensor.data_type() != ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64) {
  LOGS(logger, VERBOSE) << "The type of tensor's element data must be INT64.";
  return false;
}
```

Reply:
`DataAsSpan` does check the date type.

Reply:
> DataAsSpan does check the date type.
Cool.

Reply:
Sorry for the silly question. For some reason I wasn't getting search results for `DataAsSpan` in VS code. 🤷‍♂️

    ORT_ENFORCE(utils::IsPrimitiveDataType<T>(dtype_), "Tensor type mismatch. ",
                "T ", "!=", dtype_);